src/psyclone/psyir/nodes/dynamic_omp_task_directive.py

```diff
@@ -0,0 +1,1998 @@
+    ):
+        """
+        Evaluates a write to any Reference in the task region. This is done by
+        calling the appropriate subfunction depending on the type of the
+        Reference.
+
+        :param ref: The Reference to be evaluated.
+        :type ref: :py:class:`psyclone.psyir.nodes.Reference`
+        :param private_list: The list of private References used in this task
+                             region.
+        :type private_list: List of :py:class:`psyclone.psyir.nodes.Reference`
+        :param firstprivate_list: The list of firstprivate References used in
+                                  this task region.
+        :type firstprivate_list: List of
+                                 :py:class:`psyclone.psyir.nodes.Reference`
+        :param shared_list: The list of shared References for this task.
+        :type shared_list: List of :py:class:`psyclone.psyir.nodes.Reference`
+        :param out_list: The list of output References for this task.
+        :type out_list: List of :py:class:`psyclone.psyir.nodes.Reference`
+
+        :raises GenerationError: If a StructureReference containing multiple \
+                                 ArrayMember or ArrayOfStructuresMember as \
+                                 children is found.
+        """
+        if isinstance(ref, (ArrayReference, ArrayOfStructuresReference)):
+            self._evaluate_write_arrayref(
+                ref, private_list, firstprivate_list, shared_list, out_list
+            )
+        elif isinstance(ref, StructureReference):
+            # If the StructureReference contains an ArrayMixin then
+            # we need to treat it differently, like an arrayref, however
+            # the code to handle an arrayref is not compatible with more
+            # than one ArrayMixin child
+            array_children = ref.walk((ArrayOfStructuresMember, ArrayMember))
+            if len(array_children) > 0:
+                if len(array_children) > 1:
+                    raise GenerationError(
+                        "Doesn't support a "
+                        "StructureReference child with multiple array "
+                        "accessing members."
+                    )
+                self._evaluate_structure_with_array_reference_write(
+                    ref,
+                    array_children[0],
+                    private_list,
+                    firstprivate_list,
+                    shared_list,
+                    out_list,
+                )
+            else:
+                # This is treated the same as a Reference, but we create a
+                # Reference to the symbol to handle.
+                base_ref = Reference(ref.symbol)
+                self._evaluate_write_baseref(
+                    base_ref, private_list, shared_list, out_list
+                )
+        elif isinstance(ref, Reference):
+            self._evaluate_write_baseref(
+                ref, private_list, shared_list, out_list
+            )
```

Comment:
Please add a defensive final `else` here with an internal-error condition.

Reply:
I'm guessing this is to guard against the `lhs` of an assignment not being a `Reference` (since it seems to only be required to be a `DataNode`)?

Is there a reason the first child of `lhs` is not required for be a `Reference`? E.g. are there cases where this could still be a `CodeBlock`. I found tests doing some very strange things (e.g. `dynamo0p3_test.py::test_dynkernelargument_psyir_expression`), which seems to add a UnaryOp containing a literal as the first child of an assignment, but I couldn't work out what/why or the source code causing it so I gave up.

Reply:
It's more to guard against future changes to the code causing very unexpected things to happen (e.g. someone abusing the code or using it in a way we haven't thought of). It makes future development/debugging much easier if the code stops and tells you that something is wrong. I don't know why we don't mandate that lhs is a Reference. Could be deliberate or could be an oversight. @sergisiso might remember?